### ISSUE TITLE
pointclouds normals and features returned as None

### ISSUE THREAD

**Issue by yifita** [facebookresearch/pytorch3d#176]
If you do not know the root cause of the problem / bug, and wish someone to help you, please
post according to this template:

## 🐛 Bugs / Unexpected behaviors
When the `_points_{FORM}` is set, the methods `normals_{FORM}()` and `features_{FORM}()` return `None`, even though the normals and features are present in other forms. 
However, a more reasonable behaviour should be returning the converted normals and features.

This seems to be caused by the early exist https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/structures/pointclouds.py#L584

## Instructions To Reproduce the Issue:

```
V = torch.rand(1,5,3)
pointclouds = Pointclouds(V).to(device=device)
points_packed = pointclouds.points_packed()
pointclouds.estimate_normals(assign_to_self=True)
normals_packed = pointclouds.normals_packed()
```

`normals_packed` is `None`

**Comment by gkioxari** [comment]
Good catch! The reason is in `estimate_normals` and especially in [L890](https://github.com/facebookresearch/pytorch3d/blob/master/pytorch3d/structures/pointclouds.py#L890) where the `self._normals_packed` representation is not assigned and thus it is `None`. If you instead call `normals_list = pointclouds.normals_list()` then the output is there. Same for `normals_padded`. 

I don't think this affects the `features_X` though.

**Comment by yifita** [comment]
In general, for the heterogeneous batching structure, I find it really hard to keep track of what has been computed, what has been updated, what doesn't need recomputing, etc. 

While this feature has been promoted as a main attraction for pytorch3d, it's a little lacking compared to the expectations.

What's you advice if the pointclouds are the parameters that need optimizing. Currently, there can be lot of internal overwrite when the forms are changing from one to the other. How should one guarantee that the leaf node doesn't get "lost" during the transforms?

**Comment by gkioxari** [comment]
Thank you for the feedback. Could you be a little more concrete with what you mean? This concern is something I don't quite understand to be honest. What do you find confusing? What do you believe is overwritten? The more concrete you can be the more effective your issue is and the better chance there is for us to improve it.

**Comment by yifita** [comment]
Hi @gkioxari thank you for looking into this.

In Mesh R-CNN, a point cloud or a mesh is passed through the network once (please correctly if i'm wrong), they are the data that don't accumulate gradients.

In my case, I'd like to use the point cloud as the parameter, as in, through out the optimization/training, I have one copy of the tensors, which i keep accumulating and update values via the optimizer.

Consider the following scenario:

> initialize with list -> create packed -> changed packed -> transform to list -> update the point cloud 
list representation with the new values.

Now the list representation is no longer the initial tensors we've created, actually we probably have lost the reference to the initial tensors.

The following i'm not entirely sure:
If we repeat the above pattern in every training iteration, we are basically be creating a super long graph, starting from the initial tensor to the altered tensor in the N-th iteration. So optimizer needs to go through the entire graph to update the initial tensor.

In this case, different representations should really be just a representation pointing to the same tensor.

**Comment by gkioxari** [comment]
Hi @yifita 
I agree with @bottler. I wonder if there is something we don't quite understand in your issue. 
For us to be effective in resolving your issue if you could provide a little code snippet of an optimization loop where you feel that things are lost or not working as you expected, that would be extra useful.

**Comment by yifita** [comment]
Hi @gkioxari and @bottler thanks for the answers! I'm writing the code as I go, so I have nothing to share, only questions how to proceed :) 

I think what @bottler suggested pointed me to the right direction.
I summarize it in my own way:
In stead of using Pointclouds obj as a leaf node, construct it in every pass of the network. View it just as an method to transform the tensors.

In contrast, I was considering using them as "a container for optimizable parameters", just like the module, which is responsible for keeping the original leaf node in the graph.

Thanks! The clarification helped

**Comment by gkioxari** [comment]
Oh I see! Great! Yes, I agree with the direction you suggest and it was also the use case we had in mind when constructing the data structures. Let us know if you run into any issues!

**Comment by jcjohnson** [comment]
@yifita Yep, I think that's the right way to proceed -- if you want to directly optimize the positions of points, it's safer to re-construct a new PointClouds object at each forward pass from your optimizable parameters.

In generally it's really hard to support "a container for optimizable parameters" -- optimization algorithms generally want to mutate lists of tensors in-place, and have no way of knowing whether those tensors were really part of a `PointClouds` or `Meshes` object. So these containers would need to somehow track when any of their packed/padded/list representations were mutated in any way, and then (lazily) recompute the others. We couldn't find a good way to reliably detect any kind of in-place tensor mutation, so we chose not to support this use case. 

Even if we don't support this use case, maybe we can make it more clear in the documentation somehow.

**Comment by yifita** [comment]
> @yifita Yep, I think that's the right way to proceed -- if you want to directly optimize the positions of points, it's safer to re-construct a new PointClouds object at each forward pass from your optimizable parameters.
> 
> In generally it's really hard to support "a container for optimizable parameters" -- optimization algorithms generally want to mutate lists of tensors in-place, and have no way of knowing whether those tensors were really part of a `PointClouds` or `Meshes` object. So these containers would need to somehow track when any of their packed/padded/list representations were mutated in any way, and then (lazily) recompute the others. We couldn't find a good way to reliably detect any kind of in-place tensor mutation, so we chose not to support this use case.
> 
> Even if we don't support this use case, maybe we can make it more clear in the documentation somehow.

Agree, agree and agree.

**Comment by gkioxari** [comment]
Hi @yifita 

The commit https://github.com/facebookresearch/pytorch3d/commit/e64e0d17ef366c0d654180a0bf03a5ad21e2bf26 is fixing the issues you had with the estimated normals to update the packed computation. If your questions regarding this thread are addressed, feel free to close the issue.

**Comment by yifita** [comment]
Thanks!
